src/util/mutex.js

```diff
@@ -0,0 +1,18 @@
+/* @flow */
+
+const lockPromises: Map<any, Promise<*>> = new Map();
+
+/**
+ * Acquires a mutex lock over the given key. If the lock can't be acquired, it waits until it's available.
+ * @param key Key to get the lock for.
+ * @return {Promise.<Function>} A Promise that resolves when the lock is acquired, with the function that
+ * must be called to release the lock.
+ */
+export default (key: any): Promise<Function> => {
+  let unlockNext;
+  const willLock = new Promise((resolve) => unlockNext = resolve);
+  const lockPromise = lockPromises.get(key) || Promise.resolve();
+  const willUnlock = lockPromise.then(() => unlockNext);
+  lockPromises.set(key, lockPromise.then(() => willLock));
```

Comment:
I've based it on `await-mutex` ([code here](https://github.com/mgtitimoli/await-mutex/blob/master/src/mutex.js)). I've honestly didn't try very hard to simplify it.

I'm actually pretty surprised that a fully-functioning mutex implementation can fit in so few lines of code.